File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/async/CameraAsync.java
package vn.ditagis.com.tanhoa.qlsc.async;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.os.AsyncTask;
import android.view.LayoutInflater;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.io.ByteArrayOutputStream;

import vn.ditagis.com.tanhoa.qlsc.R;
import vn.ditagis.com.tanhoa.qlsc.entities.DApplication;


/**
 * Created by ThanLe on 4/16/2018.
 */

public class CameraAsync extends AsyncTask<byte[], Void, byte[]> {
    @SuppressLint("StaticFieldLeak")
    private AsyncResponse mDelegate;
    private DApplication mApplication;
    private Dialog mDialog;

    public interface AsyncResponse {
        void processFinish(byte[] output);
    }

    public CameraAsync(Activity activity, AsyncResponse delegate) {
        this.mDelegate = delegate;
        mDialog = new Dialog(activity);
        LayoutInflater inflater = activity.getLayoutInflater();
        LinearLayout layout = (LinearLayout) inflater.inflate(R.layout.layout_progress_dialog, null);
        TextView txtTitle = layout.findViewById(R.id.txt_progress_dialog_title);
        TextView txtMessage = layout.findViewById(R.id.txt_progress_dialog_message);
        txtTitle.setText("Vui lòng đợi");
        txtMessage.setText("Đang giảm chất lượng hình ảnh...");
        mDialog.setContentView(layout);
        mDialog.show();
    }

    private byte[] handlingCapture(byte[] bytes) {
        Bitmap decodeBitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);

        final int IMAGE_MAX_SIZE = 1000000; // 1.0MP
        int height = decodeBitmap.getHeight();
        int width = decodeBitmap.getWidth();
        double y = Math.sqrt(IMAGE_MAX_SIZE / (((double) width) / height));
        double x = (y / height) * width;
        Bitmap scaledBitmap = Bitmap.createScaledBitmap(decodeBitmap, (int) x, (int) y, true);

        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        scaledBitmap.compress(Bitmap.CompressFormat.PNG, 100, stream);
        byte[] byteArray = stream.toByteArray();

        //rotate
        Bitmap bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.length);

        Matrix matrix = new Matrix();
        matrix.postRotate(90);
        Bitmap rotatedBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        rotatedBitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream);
        bitmap.recycle();
        scaledBitmap.recycle();
        return outputStream.toByteArray();
    }

    @SuppressLint("DefaultLocale")
    @Override
    protected byte[] doInBackground(byte[]... params) {
        if (params != null && params.length > 0) {
            return handlingCapture(params[0]);
        }
        return null;
    }

    @Override
    protected void onPostExecute(byte[] result) {
        mDialog.dismiss();

        mDelegate.processFinish(result);
    }
}

File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/async/CameraAsync.kt
package vn.ditagis.com.tanhoa.qlsc.async

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.os.AsyncTask
import android.widget.LinearLayout
import android.widget.TextView

import java.io.ByteArrayOutputStream

import vn.ditagis.com.tanhoa.qlsc.R


/**
 * Created by ThanLe on 4/16/2018.
 */

class CameraAsync(activity: Activity, @field:SuppressLint("StaticFieldLeak")
private val mDelegate: AsyncResponse) : AsyncTask<ByteArray, Void, ByteArray?>() {
    //    private val mApplication: DApplication? = null
    private val mDialog: Dialog = Dialog(activity)

    interface AsyncResponse {
        fun processFinish(output: ByteArray?)
    }

    init {
        val inflater = activity.layoutInflater
        val layout = inflater.inflate(R.layout.layout_progress_dialog, null) as LinearLayout
        val txtTitle = layout.findViewById<TextView>(R.id.txt_progress_dialog_title)
        txtTitle.text = "Đang giảm chất lượng hình ảnh..."
        mDialog.setContentView(layout)
        mDialog.show()
    }

    private fun handlingCapture(bytes: ByteArray): ByteArray {
        val decodeBitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)

        val imageMaxSize = 1000000 // 1.0MP
        val height = decodeBitmap.height
        val width = decodeBitmap.width
        val y = Math.sqrt(imageMaxSize / (width.toDouble() / height))
        val x = y / height * width
        val scaledBitmap = Bitmap.createScaledBitmap(decodeBitmap, x.toInt(), y.toInt(), true)

        val stream = ByteArrayOutputStream()
        scaledBitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
        val byteArray = stream.toByteArray()

        //rotate
        val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)

        val matrix = Matrix()
        matrix.postRotate(90f)
        val rotatedBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
        val outputStream = ByteArrayOutputStream()
        rotatedBitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
        bitmap.recycle()
        scaledBitmap.recycle()
        return outputStream.toByteArray()
    }

    @SuppressLint("DefaultLocale")
    override fun doInBackground(vararg params: ByteArray): ByteArray? {
        return if (params.isNotEmpty()) {
            handlingCapture(params[0])
        } else null
    }

    override fun onPostExecute(result: ByteArray?) {
        mDialog.dismiss()

        mDelegate.processFinish(result)
    }
}